pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,381 @@
+	}
+
+	node, newProcessNode, err := at.CreateProcessNode(entry, generationType, shadowInsertion)
+	if newProcessNode && shadowInsertion {
+		return true, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	if node == nil {
+		// a process node couldn't be found or created for this event, ignore it
+		return false, err
+	}
+
+	// resolve fields
+	event.ResolveFieldsForAD()
+
+	// the count of processed events is the count of events that matched the activity dump selector = the events for
+	// which we successfully found a process activity node
+	at.Stats.processedCount[event.GetEventType()].Inc()
+
+	// insert the event based on its type
+	switch event.GetEventType() {
+	case model.ExecEventType:
+		// tag the matched rules if any
+		node.MatchedRules = model.AppendMatchedRule(node.MatchedRules, event.Rules)
+		return newProcessNode, nil
+	case model.FileOpenEventType:
+		return node.InsertFileEvent(&event.Open.File, event, generationType, at.Stats, at.shouldMergePaths, shadowInsertion)
+	case model.DNSEventType:
+		return node.InsertDNSEvent(event, generationType, at.Stats, at.DNSNames, shadowInsertion)
+	case model.BindEventType:
+		return node.InsertBindEvent(event, generationType, at.Stats, shadowInsertion)
+	case model.SyscallsEventType:
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
+		return nil, false, fmt.Errorf("process cache entry error: %s", entry.GetPathResolutionError())
+	}
+
+	// look for a ProcessActivityNode by process cookie
+	if entry.Cookie > 0 {
+		var found bool
+		node, found = at.CookieToProcessNode[entry.Cookie]
+		if found {
+			return node, false, nil
+		}
+	}
+
+	defer func() {
+		// if a node was found, and if the entry has a valid cookie, insert a cookie shortcut
+		if entry.Cookie > 0 && node != nil {
+			at.CookieToProcessNode[entry.Cookie] = node
+		}
+	}()
+
+	// find or create a ProcessActivityNode for the parent of the input ProcessCacheEntry. If the parent is a fork entry,
+	// jump immediately to the next ancestor.
+	parentNode, newProcessNode, err := at.CreateProcessNode(entry.GetNextAncestorBinary(), Snapshot, shadowInsertion)
+	if err != nil || newProcessNode && shadowInsertion {
+		return parentNode, newProcessNode, err
+	}
+
+	// if parentNode is nil, the parent of the current node is out of tree (either because the parent is null, or it
+	// doesn't match the dump tags).
+	if parentNode == nil {
+
+		// since the parent of the current entry wasn't inserted, we need to know if the current entry needs to be inserted.
+		if !at.validator.MatchesSelector(entry) {
```

Comment:
I wonder if we may want to do this check sooner (to avoid searching for a parent if the selector doesnt match), what did you think?

Reply:
So the good news here is that the cookie to process shortcut ensures that we only go up the process tree if needed, including the parent check you're talking about 🎉 
In other words, either the process as already been seen before and therefore we'll never reach the part about the parent process, or it hasn't been seen before, and then the optimal algorithm is:
- to go look for the first process that has been seen before (so that you know where to start in the tree)
- or that is at the root of the process tree and hasn't been seen before

Let's chat if it doesn't make sense :)